### PR TITLE
[1880] game end conditions, private abilities, 2p trains, force exchange rocket at b3, cross buying trains after 3 train, bcr

### PR DESCRIPTION
**Pull request by bentziaxl**
1880 implementation:
game end conditions
private abilities, 
2p trains, 
force exchange rocket at b3, 
cross buying trains after 3 train, 
bcr choose starting cert,
make corp.all_abilities return also all owning_player abilities by corp owner

### CODE REVIEW

File: migrate_game.rb
Comment (bentziaxl) [inline migrate_game.rb:55]: because of the fix to corp.all_abilities, 1889 & 18gb will have extra passed needed, when a player is prompt to use the ability

File: lib/engine/game/g_18_fl/game.rb
Comment (bentziaxl) [inline lib/engine/game/g_18_fl/game.rb:316]: the ability grants an extra tile lay. With the fix to crop.all_abilities an operator is prompt at the last operating step (after track and train buy) if they want to use the ability. Changed to the correct timing for owning player.

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (bentziaxl) [inline lib/engine/game/g_1880/step/sell_company.rb:75]: this code is very similar to buy_sell_par_shares parring choice. Should I make a common module and extract this method?

File: lib/engine/game/g_1880/round/operating.rb
Comment (bentziaxl) [inline lib/engine/game/g_1880/round/operating.rb:17]: game needs to end at the end of the turn of the corp that caused the last 8 train to go away. 
Is there a better location to place this logic? or it belongs to operating.finished?

**Comment by Voukwni** [comment]
Hello! First of all, thanks a lot for creating the game! Do you know when it will be available for play on the site?

**Comment by tobymao** [comment]
> Hello! First of all, thanks a lot for creating the game! Do you know when it will be available for play on the site?

@Voukwni this is not the appropriate discussion for that question. you can ask these kinds of questions in the 18xx slack channel

File: assets/app/view/game/corporation.rb
Comment (crericha) [inline assets/app/view/game/corporation.rb:288]: Define `ASSIGNMENT_TOKENS = {}.freeze` in base game.

Comment (crericha) [reply assets/app/view/game/corporation.rb:288]: And then revert these changes

File: assets/app/view/game/round/stock.rb
Comment (crericha) [inline assets/app/view/game/round/stock.rb:73]: what's this for?

File: assets/app/view/game/round/stock.rb
Comment (crericha) [inline assets/app/view/game/round/stock.rb:303]: Too generic. Call it something like "sell_share_text".

File: lib/engine/corporation.rb
Comment (crericha) [inline lib/engine/corporation.rb:258]: @michaeljb this code includes all owner privates with `owning_player` abilities in `all_abilities` for a corporation. Do you agree with this approach (1880 private abilities work for all our corporations) or is there another way its supposed to work?

File: lib/engine/game/g_1880/depot.rb
Comment (crericha) [inline lib/engine/game/g_1880/depot.rb:11]: super.reject {...}.min_by(&:price)

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:314]: Why reject instead of select?

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:331]: In what case would @rocket not exist?

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:337]: add a log message

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:338]: They should be able to acquire and then discard another train

> Forced exchange: ...If at this point the owner of P7 has no company with a free space for a train because of the train
limit, he can voluntarily surrender an existing train in order to create a space for the 4-train (this is
the only possible time in the game that a train can be voluntarily surrendered).

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [reply lib/engine/game/g_1880/game.rb:337]: Also export the train

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:355]: Its immediate

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:447]: Put all the non-blocking steps before the blocking

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:478]: Instead of reimplementing the method, follow what I did with CORPORATION_CLASS in base game

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:740]: Nit: More readable if `return unless trigger_sr?(train)`

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:759]: Nit: update_operators_in_saved_or

File: lib/engine/game/g_1880/round/operating.rb
Comment (crericha) [reply lib/engine/game/g_1880/round/operating.rb:17]: I think its fine

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:314]: unused and cost aren't defined in for Token. changed it to rejecting used.
I could instead select ones where token.city = nil, but this seems cleaner.

File: assets/app/view/game/round/stock.rb
Comment (bentziaxl) [reply assets/app/view/game/round/stock.rb:73]: When claiming the award for private woosong (implemented as sell_company action), I need to hide the corporations during a stock round, so that only that private is visible, and it's clear what the player needs to do (claim or pass).
Otherwise the private shows up after the corporations, and since there are 16 of them it's really not clear what the user needs to do.

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:331]: rocket closed and was used before the forced use of the rocket

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:338]: a rule detail that was lost when i wrote up the task list. Good catch

File: lib/engine/game/g_1880/depot.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/depot.rb:11]: super returns a specific train, not an array. 
` def min_depot_train
      depot_trains.min_by(&:price)
    end`

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:740]: demorgan to the rescue!

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb]: fixed the entire assign step. removed when

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb]: fixed the entire assign rocket step including this code. 
It will only export train if rocket owner owns no corps, otherwise an immediate assign followed by train drop if needed

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:478]: done!

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb]: done

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb]: fixed, i was using @rocket, instead of rocket. 
@rocket becomes null after it closes, i changed it to rocket.closed?

File: assets/app/view/game/round/stock.rb
Comment (bentziaxl) [reply assets/app/view/game/round/stock.rb:73]: this is what it looks like with hidden corps:
image

File: assets/app/view/game/round/stock.rb
Comment (crericha) [inline assets/app/view/game/round/stock.rb:303]: I meant the method "button_text" was too generic and should be called "sell_share_text" or something like that. Sorry for the confusion.

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:48]: Put this in game/base.rb instead

File: lib/engine/game/g_1880/game.rb
Comment (crericha) [inline lib/engine/game/g_1880/game.rb:750]: I'm a little slow today. Please help me understand the logic here.

The train purchased is owned by the depot or the corp with the train marker. Are you trying to determine if it was a cross-buy? If so, I don't follow how the logic is doing that.

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/sell_company.rb:53]: remove commented out code

File: assets/app/view/game/round/stock.rb
Comment (crericha) [reply assets/app/view/game/round/stock.rb:73]: It isn't clear to me from the UI what the player is supposed to do. Would this be clearer as a choice action? The 1822 cash doubler works like the latter (and happens at the end of the SR). To me, its clearer. What do you think?

https://18xx.games/game/110760?action=144

File: lib/engine/game/g_1880/step/assign.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/assign.rb:41]: Forced

File: lib/engine/game/g_1880/step/assign.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/assign.rb:47]: is still open

File: lib/engine/game/g_1880/step/buy_train.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_train.rb:68]: if entity.trains.find { |t| t.name == '2P' }

File: lib/engine/game/g_1880/step/buy_train.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_train.rb:70]: @game.can_buy_trains ? trains_to_buy : trains_to_buy.select(&:from_depot?)

Also something like can_cross_buy? would be cleared than can_buy_trains.

File: lib/engine/game/g_1880/step/company_pending_par.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/company_pending_par.rb:9]: Are the changes to this step still relevant after my parring changes? Or is there something I should review in it?

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/sell_company.rb:12]: return [] unless @game.po.owner == entity

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/sell_company.rb:19]: Why are auto actions needed? Don't we just need to prompt at the beginning of each SR or end of OR?

File: lib/engine/game/g_1880/step/company_pending_par.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/company_pending_par.rb:9]: probably not, this is for the abc/bcd choice for BCR.

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:750]: I think i broke the logic with all the revisions. It doesn't check cross buy as far as i can tell. Let me debug

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/sell_company.rb:19]: If the private's award isn't claimed by a certain phase, it's auto-claimed. The auto_actions handles that use case

File: assets/app/view/game/round/stock.rb
Comment (bentziaxl) [reply assets/app/view/game/round/stock.rb:73]: image

Comment (bentziaxl) [reply assets/app/view/game/round/stock.rb:73]: new ui, now that it's a choose and not a sell_company action

File: assets/app/view/game/round/stock.rb
Comment (bentziaxl) [reply assets/app/view/game/round/stock.rb:303]: not needed now that it's a chooce

File: lib/engine/game/g_1880/game.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/game.rb:741]: After pr merge of adding source param to buying_train!, i will fix this to:
 train.name != @depot.upcoming.first.name &&
!trains_not_triggering_sr?(train.name) &&
source == @depot

File: lib/engine/game/g_1880/step/sell_company.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/sell_company.rb:53]: removing this filr

File: lib/engine/game/g_1880/step/buy_sell_par_shares.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/buy_sell_par_shares.rb:131]: There are two entries into receive_capital:
1- when event triggers on train 3 in an OR
2- future SRs after train 3

to gain capital, a crop needs to have its fifth share sold AND not have gotten the half cap previously (to avoid double dipping)

this gain capital check is done in the method itself, so that i don't have to do that check in the event trigger! method
I could add @game.full_cap_event to the method itself, and call it without a condition in buy_sell_par_shares, but this seems more readable

File: lib/engine/game/g_1889/game.rb
Comment (bentziaxl) [inline lib/engine/game/g_1889/game.rb:259]: as discussed in slack, changing to owning_player_track. 
If a player wants to use this ability after their track step they can use it during "or_between_turns"

File: lib/engine/game/g_18_gb/entities.rb
Comment (bentziaxl) [inline lib/engine/game/g_18_gb/entities.rb:296]: I talked to 18gb implementer, there's a typo ('owning_playing' vs 'owning_player'), when isn't needed for the abilities in this game

File: lib/engine/game/g_1880/step/buy_train.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/buy_train.rb:70]: `can_cross_buy?`

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:70]: Nit: It's not really a sell, but a one-time payment

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:20]: This can go away once active is used

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:34]: "Receive #{@game.p0.name} payment now"

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:46]: "Choose whether to receive {@game.p0.name} payment"

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:53]: receives $$ from #{@company.name} one-time payment

File: lib/engine/game/g_1880/step/choose.rb
Comment (bentziaxl) [reply lib/engine/game/g_1880/step/choose.rb:70]: moved from when this was a sell_company action. renamed

File: lib/engine/game/g_1880/step/choose.rb
Comment (crericha) [inline lib/engine/game/g_1880/step/choose.rb:13]: Is this needed if there is only one active entity?

**Review by crericha — APPROVED** [review]

**Pull request merged**
